### PR TITLE
Refactor `polygon_query()`

### LINKED ISSUE #255: Refactoring `polygon_query()` and `bounding_box_query()`
TODOs:
- [x] use `sjoin` when querying by multiple polygons at the same time.
*EDIT: I implemented `sjoin` and tested the support for `MuliPolygon`. I have removed the possibility to query by multiple polygons, as the behavior that was implemented was equivalent to querying by a single `Multipolygon`.
- [x] unify APIs interface

### PR DESCRIPTION
**Pull request by LucaMarconato**
Made on top of https://github.com/scverse/spatialdata/pull/409, which should be merged first.

Closes #255.

The PR does the following:
- [x] improved comments
- [x] polygon_query doesn't transform the data but back-transforms the polygon/multipolyon. This is essential for having good performance with multiscale raster data.
- [x] unifies the `polygon_query` and `bounding_box_query` APIs. In particular
    - [x] Uses single dispatch for `polygon_query()`
    - [x] Add `sdata.query.polygon_query`
- [x] makes the functionality of `polygon_query` complete
- [x] unifies and adds more tests, in particular:
	- [x] tests `polygon_query` and `bounding_box_query` within the same test
	- [x] checks that the results are identical
    - [x] tests 3D bounding boxes with polygons
    - [x] tests multipolygons
    - [x] tests different coordinate systems via affine transformations
    - [x] tests Dask dataframes with multiple partitions.
- [x] bounding_box_query for points now doesn't reset indices
- [x] the transformation of queried elements was passed by reference instead of copy, now fixed

Minor:
- [x] `PointsModel.validate()` (which is also called by the parser) now checks that no radius value is <= 0. Added also a test for this.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -222,8 +222,7 @@ def _get_shapes() -> dict[str, GeoDataFrame]:
                 MultiPolygon(
                     [
                         Polygon(((0, 0), (0, 1), (1, 10))),
```
Comment (LucaMarconato) [inline tests/conftest.py:224]: this multipolygon was invalid (my_pol.is_valid was False), this created problems with sjoin

File: src/spatialdata/_core/query/spatial_query.py [hunk 2]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+
+    m_without_c_linear = m_without_c[:-1, :-1]
+    case = _get_case_of_bounding_box_query(m_without_c_linear, input_axes_without_c, output_axes_without_c)
+    assert case in [1, 5]
+
+    # we need to deal with the equivalent of _adjust_bounding_box_to_real_axes() to account for the fact that the
+    # intrinsic coordinate system of the points could be ('x', 'y', 'z'). The counter image of the polygon in such
+    # coordinate system could be not orthogonal to the 'z' axis; in such a case we would need to project the points to
+    # the plan in which the polygon live. Let's not handle this complexity and simply raise an error if the inverse
+    # transformation is mixing the 'z' axis with the other axes.
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:127]: I'm sorry I don't understand, when would be the case when" the inverse of the transformation is mixing the z axis" ? What does mixing means here?

File: src/spatialdata/_core/query/spatial_query.py [hunk 3]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+    output_axes_without_c
+        The axes of the query coordinate system, without the "c" axis.
+
+    """
+    from spatialdata.transformations import get_transformation
+
+    transform_to_query_space = get_transformation(element, to_coordinate_system=target_coordinate_system)
+    assert isinstance(transform_to_query_space, BaseTransformation)
+    m = _get_affine_for_element(element, transform_to_query_space)
+    input_axes_without_c = tuple([ax for ax in m.input_axes if ax != "c"])
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:181]:
```suggestion
    input_axes_without_c = tuple(ax for ax in m.input_axes if ax != "c")
```

Comment (giovp) [reply src/spatialdata/_core/query/spatial_query.py:181]: this shouldwork I think ? if yes same below

File: src/spatialdata/_core/query/spatial_query.py [hunk 4]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+    if set(axes) != set(output_axes_without_c):
+        axes_only_in_bb = set(axes) - set(output_axes_without_c)
+        axes_only_in_output = set(output_axes_without_c) - set(axes)
+
+        # let's remove from the bounding box whose axes that are not in the output axes (e.g. querying 2D points with a
+        # 3D bounding box)
+        indices_to_remove_from_bb = [axes.index(ax) for ax in axes_only_in_bb]
+        axes = tuple([ax for ax in axes if ax not in axes_only_in_bb])
+        min_coordinate = np.delete(min_coordinate, indices_to_remove_from_bb)
+        max_coordinate = np.delete(max_coordinate, indices_to_remove_from_bb)
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:207]: `min_coordinate` and `max_coordinate` are with respect to axes right? If so

```python
idx_bb = np.in1d(axes, output_axes_without_c)
idx_out = np.in1d(output_axes_without_c, axes)

min_coordinate = min_coordinate[idx_out]
max_coordinate = max_coordinate[idx_out]
```

Comment (giovp) [reply src/spatialdata/_core/query/spatial_query.py:207]: if yes, I would change `axes` to `axes_bb` and maybe `output_axes_without_c ` to `axes_out_without_x` to anyway make them consistent

Comment (giovp) [reply src/spatialdata/_core/query/spatial_query.py:207]: also, do you need to check (maybe done outside) that min/max coordinate matches in len with some axes (either bb or out)

File: src/spatialdata/_core/query/spatial_query.py [hunk 5]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+
+
+def _get_case_of_bounding_box_query(
+    m_without_c_linear: ArrayLike,
+    input_axes_without_c: tuple[str, ...],
+    output_axes_without_c: tuple[str, ...],
+) -> int:
+    """Get the case of the current bounding box query.
+
+    See https://github.com/scverse/spatialdata/pull/151 for a detailed overview of the logic of this code,
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:226]: I think that description was really cool and maybe it's worth to add it as a living document somewhere evnetually

File: src/spatialdata/_core/query/spatial_query.py [hunk 6]
```diff
@@ -619,206 +698,180 @@ def _(
...
-    points
-        Whether to filter points
+        The coordinate system of the polygon/multipolygon.
+    filter_table
+        If `True`, the table is filtered to only contain rows that are annotating regions
+        contained within the query polygon/multipolygon.
+    shapes [Deprecated]
+        This argument is now ignored and will be removed. Please filter the SpatialData object before calling this
+        function.
+    points [Deprecated]
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:739]: I don't think this is a good way to specify deprecation, I would just keep the docstrings and call the deprecation warning at functions all

File: src/spatialdata/_core/query/spatial_query.py [hunk 7]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+    transform_dimension = np.linalg.matrix_rank(m_without_c_linear)
+    transform_coordinate_length = len(output_axes_without_c)
+    data_dim = len(input_axes_without_c)
+
+    assert data_dim in [2, 3]
+    assert transform_dimension in [2, 3]
+    assert transform_coordinate_length in [2, 3]
+    assert not (data_dim == 2 and transform_dimension == 3)
+    assert not (transform_dimension == 3 and transform_coordinate_length == 2)
+    # see explanation in https://github.com/scverse/spatialdata/pull/151
```
Comment (giovp) [inline src/spatialdata/_core/query/spatial_query.py:238]: can you add a TL;DR here?

**Review by giovp — CHANGES_REQUESTED** [review]
small things around documentation

File: src/spatialdata/_core/spatialdata.py [hunk 8]
```diff
@@ -1210,24 +1211,8 @@ def bounding_box(
...
-        filter_table
-            If `True`, the table is filtered to only contain rows that are annotating regions
-            contained within the bounding box.
-
-        Returns
-        -------
-        The SpatialData object containing the requested data.
-        Elements with no valid data are omitted.
+        Please see
+        :func:`spatialdata.bounding_box_query` for the complete docstring.
```
Comment (kevinyamauchi) [inline src/spatialdata/_core/spatialdata.py:1215]: what is the reason for this change? so you don't have to maintain two docstrings? I think it's fine, but I'm curious

File: tests/models/test_models.py [hunk 9]
```diff
@@ -203,7 +203,7 @@ def test_raster_schema(
     @pytest.mark.parametrize("model", [ShapesModel])
     @pytest.mark.parametrize("path", [POLYGON_PATH, MULTIPOLYGON_PATH, POINT_PATH])
     def test_shapes_model(self, model: ShapesModel, path: Path) -> None:
-        radius = RNG.normal(size=(2,)) if path.name == "points.json" else None
+        radius = np.abs(RNG.normal(size=(2,))) if path.name == "points.json" else None
```
Comment (kevinyamauchi) [inline tests/models/test_models.py:206]: if we are going to use an RNG, we should set the seed. otherwise, test failures might be probabilistic.

File: src/spatialdata/_core/query/spatial_query.py (hunk 2 above)
Comment (kevinyamauchi) [reply src/spatialdata/_core/query/spatial_query.py:127]: This is also unclear to me. Also, what is "counter image"?

File: src/spatialdata/_core/query/spatial_query.py (hunk 5 above)
Comment (kevinyamauchi) [reply src/spatialdata/_core/query/spatial_query.py:226]: Please link to the specific comment you are referring to. That way people know where to look.

This is a private function, so I am okay with leaving it  less documented. However, I think this would be tough for a maintainer in the future to understand what's going on. Can you please add some comments throughout to explain what is going on?

File: src/spatialdata/_core/query/spatial_query.py [hunk 10]
```diff
@@ -102,6 +104,165 @@ def _get_bounding_box_corners_in_intrinsic_coordinates(
...
+            max_coordinate = np.append(max_coordinate, M)
+    return axes, min_coordinate, max_coordinate
+
+
+def _get_case_of_bounding_box_query(
+    m_without_c_linear: ArrayLike,
+    input_axes_without_c: tuple[str, ...],
+    output_axes_without_c: tuple[str, ...],
+) -> int:
+    """Get the case of the current bounding box query.
```
Comment (kevinyamauchi) [inline src/spatialdata/_core/query/spatial_query.py:224]: Can you please clarify what you mean by "case" here? I see below that you intend to link to a larger discussion. However, people need to be able to understand at least at a high level what the function does.

File: src/spatialdata/_core/query/spatial_query.py [hunk 11]
```diff
@@ -585,9 +657,14 @@ def _(
         min_coordinate=min_coordinate,
         max_coordinate=max_coordinate,
     )
-    if bounding_box_mask.sum() == 0:
+    bounding_box_mask = np.where(bounding_box_mask.compute())[0]
```
Comment (kevinyamauchi) [inline src/spatialdata/_core/query/spatial_query.py:660]: nitpick, but these are indices, right? I think a mask is usually a bool.

**Review by kevinyamauchi — COMMENTED** [review]
Thanks, @LucaMarconato . I think this looks pretty good. I've left some comments below, mostly related to code clarity.

I am okay with the refactors in `test_spatial_query`. I see that they likely reduce the amount of code. However, I think this is at the cost of being able to quickly understand the intention of the test. If that's the trade-off we want, I am fine with it. I would add some comments and docstrings to make it easier to understand though.

File: src/spatialdata/_core/query/spatial_query.py (hunk 11 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:660]: true, I changed this

File: src/spatialdata/_core/query/spatial_query.py (hunk 2 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:127]: I will rewrite the comment in a clearer way. Here "counter image" is a bad translation of the Italian "controimmagine". I checked and it's called "preimage" in English.

Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:127]: Done

File: src/spatialdata/_core/query/spatial_query.py (hunk 3 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py]: true, replaced

Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py]: true, replaced

File: src/spatialdata/_core/query/spatial_query.py (hunk 10 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:233]: Improved the first line and referring to the comments for a TL;DR

File: src/spatialdata/_core/query/spatial_query.py (hunk 5 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:235]: I lnked the specific comment, also we can use GitHub as a living document as the core devs can edit the comments of other people.

File: src/spatialdata/_core/query/spatial_query.py (hunk 7 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:247]: Added.

File: src/spatialdata/_core/query/spatial_query.py (hunk 6 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py:771]: Done.

File: src/spatialdata/_core/spatialdata.py (hunk 8 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:1215]: Exactly, I would like to have it only in one place because it's easy to forget to update the other (happened). Is there any Sphinx trick we can use here? Asking also @giovp. For this PR I'll keep it like this because I'd like to merge some other PRs dependent on this.

Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:1215]: I see also torch has something similar to what is in this PR: https://pytorch.org/docs/stable/generated/torch.Tensor.sum.html

File: tests/models/test_models.py (hunk 9 above)
Comment (LucaMarconato) [reply tests/models/test_models.py:206]: true, added

File: src/spatialdata/_core/query/spatial_query.py (hunk 4 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py]: > min_coordinate and max_coordinate are with respect to axes right? If so
>
> idx_bb = np.in1d(axes, output_axes_without_c)  
> idx_out = np.in1d(output_axes_without_c, axes)  

this still returns a boolean mask, so it's basically equivalent to what I wrote; I'll keep the old code.

min_coordinate = min_coordinate[idx_out]  
max_coordinate = max_coordinate[idx_out]

Comment (LucaMarconato) [reply src/spatialdata/_core/query/spatial_query.py]: yes, we check that it matches with bb, and we do it `BoundingBoxRequest.__post_init__()`
